cpp/unreal_plugins/SpEngine/Source/SpEngine/EngineService.h

```diff
@@ -0,0 +1,175 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#pragma once
+
+#include <atomic>
+#include <concepts> // std::same_as
+#include <future>   // std::promise, std::future
+#include <string>
+
+#include <Delegates/IDelegateInstance.h> // FDelegateHandle
+#include <Misc/CoreDelegates.h>
+
+#include "SpCore/Assert.h"
+#include "SpCore/Log.h"
+#include "SpEngine/WorkQueue.h"
+
+template <typename TBasicEntryPointBinder>
+concept CBasicEntryPointBinder = requires(TBasicEntryPointBinder basic_entry_point_binder) {
+    { basic_entry_point_binder.bind("", []() -> void {}) } -> std::same_as<void>;
+};
+
+template <typename TEntryPointBinder>
+concept CEntryPointBinder = requires(TEntryPointBinder entry_point_binder) {
+    { entry_point_binder.bind_func("", "", []() -> void {}) } -> std::same_as<void>;
+    { entry_point_binder.bind_func_wrapped("", "", []() -> void {}) } -> std::same_as<void>;
+};
+
+// Different possible frame states for thread synchronization
+enum class FrameState
+{
+    Idle,
+    RequestPreTick,
+    ExecutingPreTick,
+    ExecutingTick,
+    ExecutingPostTick,
+};
+
+template <CBasicEntryPointBinder TBasicEntryPointBinder>
+class EngineService {
+public:
+    EngineService() = delete;
```

Comment:
If the default constructor wasn't deleted, could we use an object instead of a pointer for `EngineService`?

Reply:
No, we cannot have an object of `EngineService`. Since `EngineService` has a member variable `WorkQueue` which has a member variable `boost::asio::io::context` whose copy assignment operator is deleted, we cannot do an operation such as 
```cpp
// SpEngine.h
EngineService<rpc::server> engine_service_;

// SpEngine.cpp
rpc_server_ = ...;
engine_service_ = EngineService<rpc::server>(rpc_server_.get());
```

Reply:
Ah that's right. I'm fine with all the services being pointers, but I still think the RPC server can be an object.